### PR TITLE
Corgi abl

### PR DESCRIPTION
**Pull request by TylerZeroMaster**
Fixes openstax/corgi#496

## Frontend changes:

- Adds a new interface for approving books
   - When approving a book, you can select the last code version or the worker version
- Adds factory functions to frontend unit tests so that developers do not need to create test data from scratch for every test
- Updates the CORGI UI tests to support interface changes
- Highlights jobs that built books contained in the ABL
- Adds more information to job details dialog (repository name & job type)

## Backend changes:
- Updates the integration tests to ensure the new endpoint works as intended
- Adds new fixtures to backend unit tests that support snapshot testing of sqlalchemy operations
- Add new endpoints to modify and read the approved books list (ABL).
   - The new endpoint replaces the old `/api/abl/` endpoint and supports the following functionality:
      - Get all ABL entries, or a subset based on consumer (i.e. "REX"), repo_name, version (commit sha), or code_version
         - Entries are returned as a flat list of books. Each entry contains uuid, commit_sha, code_version, consumer, created_at, committed_at, repository_name, and slug
      - Add ABL entries
        - For REX consumer, adding a new entry keeps the entry present on their release.json and at most one newer version
        - For all other consumers, adding a new entry only keeps the version being added

## Screenshots

Web gets the new approve interface
Screenshot 2024-03-08 at 10 13 15 AM

The new approve interface only appears for web

Screenshot 2024-03-08 at 10 15 01 AM

Screenshot 2024-03-08 at 10 15 31 AM

Screenshot 2024-03-08 at 10 17 06 AM

### CODE REVIEW

File: frontend/src/components/DetailsDialog.svelte [hunk 1]
```diff
@@ -28,37 +29,41 @@
...
     <Content>
       {#if selectedJob.status.name === "completed"}
-        {#each selectedJob.artifact_urls as artifact}
-          <a href={artifact.url} target="_blank" rel="noreferrer"
-            >{artifact.slug}</a
-          >
-          <br />
-        {/each}
+        <BuildArtifacts {selectedJob}></BuildArtifacts>
+        {#if selectedJob.job_type.name === "git-web-hosting-preview"}
```
Comment (philschatz) [inline frontend/src/components/DetailsDialog.svelte:57]: (note to self) This seems to be the place in the UI that decides whether a book can be approved... but it seems like the BE allows other options? (maybe allow other options here or remove other options there?)

File: frontend/src/ts/abl.ts [hunk 2]
```diff
@@ -0,0 +1,77 @@
+import { RequireAuth } from "./fetch-utils";
+import type { Job, Book, ApprovedBookWithDate } from "./types";
+import { handleError } from "./utils";
+import { parseDateTimeAsUTC } from "./utils";
+
+export async function newABLentry(job: Job, codeVersion: string) {
+  const guessConsumer = (b: Book) =>
+    b.slug.match(/-(ancillaries|ancillary)$/) != null ? "ancillary" : "REX";
```
Comment (philschatz) [inline frontend/src/ts/abl.ts:8]: :sunglasses:

**Review by philschatz — APPROVED** [review]
:wave:

File: frontend/src/components/DetailsDialog.svelte (hunk 1 above)
Comment (TylerZeroMaster) [reply frontend/src/components/DetailsDialog.svelte:57]: This seems to be the place in the UI that decides whether a ~book can be approved~ job can be used to get a list of books to approve.

File: backend/app/tests/unit/test_abl.py [hunk 3]
```diff
@@ -0,0 +1,163 @@
...
+                ),
+            ],
+            {"b": {"defaultVersion": "b"}, "c": {"defaultVersion": "c"}},
+        ],
+    ],
+)
+async def test_add_new_entries_rex(
+    to_add, to_keep, mock_session, mock_http_client, snapshot
+):
+    def mock_database_logic(session_obj):
```
[the remark this answers is not in the document]
Comment (TylerZeroMaster) [reply backend/app/tests/unit/test_abl.py:91]: @philschatz In the past, I had suggested using sqlite for unit tests and was told "unit tests should require as few dependencies as possible" and that I should "just check to make sure that the correct inputs are sent to sqlalchemy". 

It sounds like you are talking about integration testing, which is handled separately.

File: README.md [hunk 4]
```diff
@@ -360,6 +374,13 @@ This section describes the data model used by CORGI. For more information
 about how to read this diagram, see [Crow's foot notation](https://en.wikipedia.org/wiki/Entity%E2%80%93relationship_model#Crow's_foot_notation).
 ```mermaid
 erDiagram
+    ApprovedBook {
+        int book_id
+        datetime created_at
+        datetime updated_at
```
[the remark this answers is not in the document]
Comment (TylerZeroMaster) [reply README.md:380]:
```mermaid
erDiagram
    ApprovedBook {
        int book_id
        datetime created_at
        datetime updated_at
        int consumer_id
        int code_version_id
    }
    Book {
        int id
        str slug
        str uuid
        str style
        int edition
        int commit_id
    }
    Commit {
        int id
        str sha
        datetime timestamp
        int repository_id
    }
    Book ||--|{ ApprovedBook : ""
    Commit ||--|{ Book : ""
```

- A commit can have many books, but a book only belongs to one commit.
- A book can have many approved book entries, but an approved book entry can only have one book.
- An approved book entry can only have one book which can only have belong to one commit.

File: backend/app/tests/unit/test_abl.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (TylerZeroMaster) [reply backend/app/tests/unit/test_abl.py:92]: As for conditional mocking of database queries: there is a package called `mock-alchemy` that does the same thing in a more complete package. Unfortunately, I have not been able to get this package working with the version of SQLAlchemy that we are using.

**Pull request merged**